### PR TITLE
[hail] experimental support for encoders and decoders

### PR DESCRIPTION
**Pull request by danking**
I rolled Jackie, but @tpoterba you should check you're happy with the code structure.

Adds two hail experimental functions, `encode` and `decode`, which provide a python interface to the encoders and decoders in `RowStore.scala`:

```
In [16]: import hail as hl 
    ...: import hail.experimental as hle                                                                                                                       

In [17]: hle.encode(hl.literal(1))                                                                                                                             
Out[17]: ('Tuple[Int32]', b'\x07\x00\x00\x00\x02\x00\x00\x00 \x00\x01')

In [18]: hle.encode(hl.literal(1), 'defaultUncompressed')                                                                                                      
Out[18]: ('Tuple[Int32]', b'\x05\x00\x00\x00\x00\x01\x00\x00\x00')

In [19]: hle.encode(hl.literal(1), 'unblockedUncompressed')                                                                                                    
Out[19]: ('Tuple[Int32]', b'\x00\x01\x00\x00\x00')

In [20]: hle.encode(hl.literal([1,2,3,4,5,6]))                                                                                                                 
Out[20]: 
('Tuple[Array[Int32]]',
 b'\x0e\x00\x00\x00\t\x00\x00\x00\x90\x00\x06\x00\x01\x02\x03\x04\x05\x06')

In [21]: hle.encode(hl.literal(1), 'defaultUncompressed')                                                                                                      
Out[21]: ('Tuple[Int32]', b'\x05\x00\x00\x00\x00\x01\x00\x00\x00')

In [22]: hle.encode(hl.literal(1), 'unblockedUncompressed')                                                                                                    
Out[22]: ('Tuple[Int32]', b'\x00\x01\x00\x00\x00')

In [23]: v = hl.struct(x=1) 
    ...: hle.decode(v.dtype, *hle.encode(v))                                                                                                                   
Out[23]: Struct(x=1)
```

### CODE REVIEW

File: hail/src/main/scala/is/hail/backend/Backend.scala [hunk 1]
```diff
@@ -94,5 +96,36 @@ abstract class Backend {
...
+    val ir = lower(ir0, None, false)
+
+    if (!Compilable(ir))
+      throw new LowererUnsupportedOperation(s"lowered to uncompilable IR: ${Pretty(ir)}")
+    if (ir.typ == TVoid)
+      throw new LowererUnsupportedOperation(s"lowered to unit-valued IR: ${Pretty(ir)}")
+
+    Region.scoped { region =>
+      val (pt: PTuple, f) = Compile[Long](MakeTuple(FastSeq(ir)))
+      (pt.parsableString(), codec.encode(pt, region, f(0)(region)))
```
Comment (jigold) [inline hail/src/main/scala/is/hail/backend/Backend.scala]: What's the index 0 for? I tried to tell from the code.

File: hail/src/main/scala/is/hail/io/RowStore.scala [hunk 2]
```diff
@@ -177,13 +183,23 @@ trait CodecSpec extends Serializable {
...
 
+  def encode(t: PType, region: Region, offset: Long): Array[Byte] = {
+    val baos = new ByteArrayOutputStream()
+    using(buildEncoder(t)(baos))(_.writeRegionValue(region, offset))
+    baos.toByteArray()
+  }
+
+  def decode(t: PType, bytes: Array[Byte], region: Region): Long = {
+    val bais = new ByteArrayInputStream(bytes)
+    buildDecoder(t, t)(bais).readRegionValue(region)
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/RowStore.scala]: When would `t` and `requestedType` be different? Should there be an interface that only includes one type like there is with `buildEncoder`?

File: hail/src/main/scala/is/hail/backend/Backend.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/backend/Backend.scala]: partition index, used for randomness

File: hail/src/main/scala/is/hail/io/RowStore.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/io/RowStore.scala]: encoder needs both as well, actually.

These two differ when we're decoding a Table from disk -- the requested type is often a (smaller) supertype of the full type.

**Pull request merged**